server/src/test/java/org/elasticsearch/snapshots/RestoreServiceTests.java

```diff
@@ -109,4 +128,73 @@ public void testPrefixNotChanged() {
+        final Set<String> pendingRefreshes = new HashSet<>();
+        final List<Runnable> finalAssertions = new ArrayList<>();
+        while (repositories.size() < repositoryCount) {
+            final String repositoryName = randomAlphaOfLength(10);
+            switch (between(1, 3)) {
+                case 1:
+                    final Repository notBlobStoreRepo = mock(Repository.class);
+                    repositories.put(repositoryName, notBlobStoreRepo);
+                    finalAssertions.add(() -> verifyZeroInteractions(notBlobStoreRepo));
+                    break;
+                case 2:
+                    final Repository freshBlobStoreRepo = mock(BlobStoreRepository.class);
+                    repositories.put(repositoryName, freshBlobStoreRepo);
+                    when(freshBlobStoreRepo.getMetadata()).thenReturn(
+                            new RepositoryMetadata(
+                                    repositoryName,
+                                    randomAlphaOfLength(3),
+                                    Settings.EMPTY
+                            ).withUuid(UUIDs.randomBase64UUID()));
+                    doThrow(new AssertionError("repo UUID already known")).when(freshBlobStoreRepo).getRepositoryData(any());
+                    break;
+                case 3:
+                    final Repository staleBlobStoreRepo = mock(BlobStoreRepository.class);
+                    repositories.put(repositoryName, staleBlobStoreRepo);
+                    pendingRefreshes.add(repositoryName);
+                    when(staleBlobStoreRepo.getMetadata()).thenReturn(
+                            new RepositoryMetadata(
+                                    repositoryName,
+                                    randomAlphaOfLength(3),
+                                    Settings.EMPTY
+                            ));
+                    doAnswer(invocationOnMock -> {
+                        assertTrue(pendingRefreshes.remove(repositoryName));
+                        //noinspection unchecked
+                        ActionListener<RepositoryData> repositoryDataListener
+                                = (ActionListener<RepositoryData>) invocationOnMock.getArguments()[0];
+                        if (randomBoolean()) {
+                            repositoryDataListener.onResponse(null);
+                        } else {
+                            repositoryDataListener.onFailure(new Exception("simulated"));
+                        }
+                        return null;
+                    }).when(staleBlobStoreRepo).getRepositoryData(any());
+                    break;
+            }
+        }
+
+        final RepositoriesService repositoriesService = mock(RepositoriesService.class);
+        when(repositoriesService.getRepositories()).thenReturn(repositories);
+        RestoreService.refreshRepositoryUuids(true, repositoriesService, listener);
+        assertNull(listener.actionGet(0L));
```

Comment:
NIT: `.get()` to have a stack trace on failure? (I guess failure is impossible, here but just in case)